README.md
Sherlock
========

_Sherlock_ is a PHP (>=5.3.9) client for [ElasticSearch](http://www.elasticsearch.org/).  _Sherlock_ can be used to search and manage ElasticSearch clusters.

Features
--------

 - One-to-one mapping with ElasticSearch's API and query DSL.
 - Option to use powerful ORM-like interface, associative arrays or raw JSON
 - Concurrent request streaming with RollingCurl and cURL multi-handle
 - Autodetection of cluster nodes and subsequent round-robin requests to the entire cluster
 - Configurable logging capabilities

Disclaimer
--------
Sherlock should be considered **Alpha** status.  Use in production at your own risk!  The current 0.1 branch is in maintenance mode - I'm only fixing bugs and accepting PRs.

The new 0.2 branch that is coming *will* break backwards compatibility in places.  The new 0.2 branch, however, should be a lot easier to maintain and I'll likely be promoting it to true 1.0 status soon after release (along with associated SemVer, backwards compatibility, etc).

Resources
---------------
 - Read the [Quickstart Guide](http://sherlockphp.com/quickstart/)
 - Read the [Full Documentation](http://sherlockphp.com/documentation/)
 - [Roadmap for Sherlock](Roadmap.md)
 - [Changelog](Changelog.md)

Installation via Composer
-------------------------
The recommended method to install _Sherlock_ is through [Composer](http://getcomposer.org).

1. Add ``sherlock/sherlock`` as a dependency in your project's ``composer.json`` file:

        {
            "require": {
                "sherlock/sherlock": "~0.1.0"
            }
        }

   *Note*: If you would like to use the latest stable development version, specify ``dev-master`` in your composer file instead of ``0.1.*``.  Master is kept at the most recent, stable development version but may not be tagged and pushed to Packagist yet.  Unstable dev versions are kept secluded in the ``develop`` branch.

2. Download and install Composer:

        curl -s http://getcomposer.org/installer | php

3. Install your dependencies:

        php composer.phar install

4. Require Composer's autoloader

    Composer also prepares an autoload file that's capable of autoloading all of the classes in any of the libraries that it downloads. To use it, just add the following line to your code's bootstrap process:

        <?php
        require 'vendor/autoload.php';

You can find out more on how to install Composer, configure autoloading, and other best-practices for defining dependencies at [getcomposer.org](http://getcomposer.org).


Manual Installation
-------------------
_Sherlock_ can be installed even if you don't use Composer.  Download Sherlock and include the following in your index.php or equivalent

```php
        <?php
        require 'Sherlock/Sherlock.php';
        \Sherlock\Sherlock::registerAutoloader();
```

Usage
-----
The library interface is still under flux...this section will be updated once _Sherlock_ has been fleshed out a bit more.

```php
   require 'vendor/autoload.php';
   use \Sherlock\Sherlock;

   //The Sherlock object manages cluster state, builds queries, etc
   $sherlock = new Sherlock();

   //Add a node to our cluster.  Sherlock can optionally autodetect nodes given one starting seed
   $sherlock->addNode('localhost', 9200);

   //Build a new search request
   $request = $sherlock->search();

   //Set the index, type and from/to parameters of the request.
   //The query is at the end of the chain, although it could be placed anywhere
   $request->index("test")
            ->type("tweet")
            ->from(0)
            ->size(10)
            ->query(Sherlock::query()->Term()->field("message")
                                              ->term("ElasticSearch"));

   //Execute the search and return results
   $response = $request->execute();

   echo "Took: ".$response->took."\r\n";
   echo "Number of Hits: ".count($response)."\r\n";

   //Iterate over the hits and print out some data
   foreach($response as $hit)
   {
      echo $hit['score'].' - '.$hit['source']['message']."\r\n";
   }


   //Let's try a more advanced query now.
   //Each section is it's own variable to help show how everything fits together
   $must = Sherlock::query()->Term()->field("message")
                                     ->term("ElasticSearch");

   $should = Sherlock::query()->Match()->field("author")
                                        ->query("Zachary Tong")
                                        ->boost(2.5);

   $must_not = Sherlock::query()->Term()->field("message")
                                           ->term("Solr");

   $bool = Sherlock::query()->Bool->must($must)
                                   ->should($should)
                                   ->must_not($must_not);
   $request->query($bool);
   $request->execute();

```

Other types of queries
----------------------
You can use _Sherlock_ with every type of query listed in the elasticsearch docs.
E.g. if you'd like to use a _fuzzy like this (flt)_ query, you can build your query like this:

```php
	$sherlock = new Sherlock();
    $sherlock->addNode('localhost', 9200);
    $request = $sherlock->search();

	$request->index('test')
			->type('tweet')
			->query(Sherlock::queryBuilder()
				->FuzzyLikeThis()
				->fields( array('description', 'tags', 'name') )
				->like_text( $query )
				->min_similarity( 0.6 )
			);

	$response = $request->execute();
```

Filters
-------
Building filters is identical to building queries, but requires the use of _filterBuilder()_ instead of _queryBuilder()_.
Again, a simple example would be:

```php
    $request->index('test')
		->type('tweet')
		->filter(Sherlock::filterBuilder()
			->Term()
			->field($type)
			->term($value)
		);
	
	$response = $request->execute();
```



Non-ORM style
-------------
Not a fan of ORM style construction?  Don't worry, _Sherlock_ supports "raw" associative arrays

```php
    //Build a new search request
    $request = $sherlock->search();

    //We can compose queries using hashmaps instead of the ORM.
    $manualData = array("field" => "field1", "term" => "town");

    $request->query(Sherlock::query()->Term($manualData));

```

Need to consume and use raw JSON?  No problem
```php
    //Build a new search request
    $request = $sherlock->search();

    //We can compose queries using hashmaps instead of the ORM.
    $json = '{ "term" : { "field1" : "town" } }';

    $request->query(Sherlock::query()->Raw($json));

```

(There will be a RawQuery method soon, that lets you construct entirely arbitrary queries with arrays or JSON)

For more examples check out the [Quickstart Guide](http://sherlockphp.com/quickstart.html)


Philosophy
----------
_Sherlock_ aims to continue the precendent set by ElasticSearch: work out of the box with minimal configuration and provide a simple interface.

_Sherlock's_ API uses a "fluent" interface that relies heavily on method chaining and type hinting for brevity.  The developer should never need to stop and look up a class name to instantiate, or dig through docs to remember which class accepts which arguments.

Secondary to the interface comes _Sherlock_ developer sanity: reduce code as much as possible.  Rather than write a million getter/setter functions to expose ElasticSearch's various parameters, _Sherlock_ relies heavily upon templates, auto-generated class stubs, magic methods and PHPdoc.


Changelog.md
Changelog
========

Master
------

0.1.14
------
 - Fix: Facets now accepts arrays and inline arguments

0.1.14
------
 - Fix: OrFilter now works appropriately if _cache is specified with multiple filters/queries
 - Fix: AndFilter now behaves like OrFilter

0.1.13
------
 - Fix: Fix _Analyzer mapping
 - Change: RawRequest added, RawQuery "downgraded" to affecting only "query: {}" portion of query, RawFilter + RawFacet added
 - Fix: Bool Filter/Query fixed when a single object is provided
 - Fix: Throw a new DocumentMissingException when attempting to delete a nonexistent document
 - Fix: FilterQuery now correctly converts child filter data to json
 - Feature: Add boost param to Term Query
 - Feature: Add fields param to Search request
 - Feature: Add GeoPoint mappings
 - In-progress: Refactor components to use static methods
 - In-progress: Refactor components to set only arguments that have been sepcified (no more defaults)
 - Fix: General refactoring
 - Fix: CustomFilterScore now supports the boost parameter
 - Addition: custom CodeSniffer code standard now included in repo for contributors to use

0.1.12
------
 - Feature: Performance enhancements to RollingCurl when the pending request buffer is large
 - Feature: Add filter_facets to facets
 - Feature: Add Update capability to indexing operations
 - Feature: Add _Analyzer mapping support

0.1.11
------
 - Fix: Mappings on index creation were not working properly

0.1.10
------
 - Fix: Search requests were using malformed JSON
 - Fix: All requests lacked appropriate Content-Type headers, leading to HTTP requests with url-encoding specified

0.1.9
-----
 - Fix: Relax the version requirement for Event-Dispatcher to ~2.1

0.1.8
-----
 - Internal: Replace Guzzle with RollingCurl (also [Fixes #20](https://github.com/polyfractal/sherlock/issues/20))
 - Feature: Batch document indexing. [Closes #24](https://github.com/polyfractal/sherlock/issues/24)
 - Feature: Some of the more commone ES errors are now returned as Exceptions. [Closes #21](https://github.com/polyfractal/sherlock/issues/21)
 - Fix: Raw query now is 100% "raw", rather than wrapping inside a "query". [Closes #23](https://github.com/polyfractal/sherlock/pull/23)

0.1.7
-----
 - Breaking Change: Sherlock static builder functions refactored with consistent naming scheme [Fixes #5](https://github.com/polyfractal/sherlock/issues/5)
 - Internal: Refactor composeFinalQuery to use assoc arrays instead of strings [Fixes #13](https://github.com/polyfractal/sherlock/issues/13)
 - Feature: Add support for all facets (Terms, Range, Histogram, DateHistogram, Filter, Query, Statistical, TermsStats, Geodistance) [Closes #16](https://github.com/polyfractal/sherlock/issues/16)

0.1.6
-----
 - Fix: Sort ordering fixed, uses "order" instead of "sort_order" [Fixes #14](https://github.com/polyfractal/sherlock/pull/14)
 
0.1.5
-----
 - Fix: Change default logging state to disabled [474fa95](https://github.com/polyfractal/sherlock/commit/474fa957c61b550fa043315757a4e279179dc0d8)
 - Fix: Integrate Symfony EventDispatcher to decouple nodes from Requests. [Fixes #10](https://github.com/polyfractal/sherlock/issues/10)
 - Feature: Enable custom logging handlers [Usage](https://github.com/polyfractal/sherlock/issues/12#issuecomment-14682664) [474fa95](https://github.com/polyfractal/sherlock/commit/474fa957c61b550fa043315757a4e279179dc0d8)

0.1.4
-----
 - Fix: Updated Analog dependency to remove a logging bug [4c3c69e](https://github.com/polyfractal/sherlock/commit/4c3c69e59365784e70c2ec8d0d83bc8d5a060fda)

0.1.3
------
 - Feature: Enable sorting on Search Requests
 - Fix: From/Size fixed
 - Fix: Match query now allows 'type' to be set
 - Fix: Default fuzziness for Match query set to 'null'
 - Fix: Default fuzziness for MultiMatch query set to 'null'





src/Sherlock/common/Cluster.php
<?php
/**
 * User: Zachary Tong
 * Date: 3/10/13
 * Time: 11:23 AM
 */

namespace Sherlock\common;

use Sherlock\common\events\RequestEvent;
use Sherlock\common\exceptions\RuntimeException;
use Symfony\Component\EventDispatcher\EventDispatcher;

/**
 * Class Cluster
 * @package Sherlock\common
 */
class Cluster
{
    /**
     * @var array
     */
    private $nodes = array();

    /**
     * @var EventDispatcher
     */
    private $dispatcher;


    /**
     * @param EventDispatcher $dispatcher
     */
    public function __construct(EventDispatcher $dispatcher)
    {
        $this->dispatcher = $dispatcher;
    }


    /**
     * @param  string                              $host
     * @param  int                                 $port
     * @param  bool                                $autodetect
     *
     * @throws exceptions\BadMethodCallException
     * @throws exceptions\InvalidArgumentException
     */
    public function addNode($host, $port, $autodetect = true)
    {
        if (!isset($host)) {
            throw new exceptions\BadMethodCallException("A server address must be provided when adding a node.");
        }

        if (!is_numeric($port)) {
            throw new exceptions\InvalidArgumentException("Port argument must be a number");
        }

        $this->nodes[$host] = array('host' => $host, 'port' => $port);
    }


    /**
     * Autodect various cluster properties
     */
    public function autodetect()
    {
        $this->autodetect_parseNodes();
    }


    /**
     * Triggered just prior to a request being executed
     * Inject a random node into the Request object
     *
     * @param  RequestEvent                $event
     *
     * @throws exceptions\RuntimeException
     */
    public function onRequestExecute(RequestEvent $event)
    {
        $request = $event->getRequest();

        //Make sure we have some nodes to choose from
        if (count($this->nodes) === 0) {
            throw new RuntimeException("No nodes in cluster, request failed");
        }

        //Choose a random node
        $request->node = $this->nodes[array_rand($this->nodes)];

    }


    /**
     * Autodetect the nodes in this cluster through Cluster State API
     */
    private function autodetect_parseNodes()
    {
        foreach ($this->nodes as $node) {
            try {
                $client   = new Client('http://' . $node['host'] . ':' . $node['port']);
                $request  = $client->get('/_nodes/http');
                $response = $request->send()->json();

                foreach ($response['nodes'] as $newNode) {

                    //we don't want http-inaccessible nodes
                    if (!isset($newNode['http_address'])) {
                        continue;
                    }

                    preg_match('/inet\[\/([0-9\.]+):([0-9]+)\]/i', $newNode['http_address'], $match);

                    $tNode = array('host' => $match[1], 'port' => $match[2]);

                    //use host as key so that we don't add duplicates
                    $this->nodes[$match[1]] = $tNode;

                }

                //we have the complete node list, no need to keep checking
                break;

            } catch (\Guzzle\Http\Exception\BadResponseException $e) {
                //error with this node, continue onto the next one
            }
        }
    }
}
